### PR TITLE
Use MaybeUninit more properly in streamer::batch_send()

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
`batch_send()` has undefined behavior. Some uninitialized data is marked as initialized with `assume_init()` even though it hasn't actually been initialized:
https://github.com/anza-xyz/agave/blob/10bebfaf422b8e09409cc5b56a35ec40defa27aa/streamer/src/sendmmsg.rs#L138-L140

This is problematic since the [iovec](https://docs.rs/libc/latest/libc/struct.iovec.html) struct has a pointer:
```rust
#[repr(C)]
pub struct iovec {
    pub iov_base: *mut c_void,
    pub iov_len: size_t,
}
```
With rust 1.82, this made the validator start crashing:
```
kernel: traps: solRspndrGossip[1514252] trap invalid opcode ip:561a11d83bb7 sp:7eb14c3a4830 error:0
sol.service: Main process exited, code=killed, status=4/ILL
```

#### Summary of Changes
- Use `MaybeUninit` interface to initialize `iovec`'s
- Use `MaybeUninit` interface to initialize `mmsghdr`'s
- Write every field in `mmsghdr` so that we no longer need to request zeroed data
- Use `MaybeUninit` interface to initialize `sockaddr_storage`'s

### CODE REVIEW

**Comment by steviez** [comment]
Looks like we might have the same problem here; will tack another commit onto this PR as I think the solution should be nearly identical:
https://github.com/anza-xyz/agave/blob/10bebfaf422b8e09409cc5b56a35ec40defa27aa/streamer/src/recvmmsg.rs#L88-L89

**Comment by steviez** [comment]
Two things as I review this real quick:
- Should probably refactor `addrs` and `hdrs` to use `MaybeUninit` API as well
- `recv_mmsg()` has similar issues as `batch_send()`; I prefer smaller PR's so am initially inclined to do that separately but TBD

File: streamer/src/sendmmsg.rs [hunk 1]
```diff
@@ -135,14 +146,20 @@ where
     T: AsRef<[u8]>,
 {
     let size = packets.len();
-    #[allow(clippy::uninit_assumed_init)]
-    let iovec = std::mem::MaybeUninit::<iovec>::uninit();
-    let mut iovs = vec![unsafe { iovec.assume_init() }; size];
-    let mut addrs = vec![unsafe { std::mem::zeroed() }; size];
-    let mut hdrs = vec![unsafe { std::mem::zeroed() }; size];
+    let mut iovs = vec![MaybeUninit::<iovec>::uninit(); size];
+    let mut addrs = vec![MaybeUninit::<sockaddr_storage>::zeroed(); size];
```
Comment (alessandrod) [inline streamer/src/sendmmsg.rs]: nit: there's no need to explicitly set the type, use MaybeUninit::uninit() and MaybeUninit::zeroed()

File: streamer/src/sendmmsg.rs [hunk 2]
```diff
@@ -135,14 +146,20 @@ where
...
+    let mut hdrs = vec![MaybeUninit::<mmsghdr>::uninit(); size];
     for ((pkt, dest), hdr, iov, addr) in izip!(packets, &mut hdrs, &mut iovs, &mut addrs) {
         mmsghdr_for_packet(pkt.as_ref(), dest.borrow(), iov, addr, hdr);
     }
+    // mmsghdr_for_packet() performs initialization so we can safely transmute
+    // the Vecs to their initialized counterparts
+    let _iovs = unsafe { std::mem::transmute::<Vec<MaybeUninit<iovec>>, Vec<iovec>>(iovs) };
```
Comment (alessandrod) [inline streamer/src/sendmmsg.rs]: nit: import std::{mem, ptr} and use mem::transmute and ptr::write

**Review by alessandrod — DISMISSED** [review]

File: streamer/src/sendmmsg.rs (hunk 2 above)
Comment (steviez) [reply streamer/src/sendmmsg.rs]: Done in https://github.com/anza-xyz/agave/pull/3325/commits/630c1dbb61e36704cd5095148b07b53ec48ef144

I did the same for another few instances for the sake of keeping use across the file consistent

File: streamer/src/sendmmsg.rs (hunk 1 above)
Comment (steviez) [reply streamer/src/sendmmsg.rs]: Done in https://github.com/anza-xyz/agave/pull/3325/commits/67ff531cbb56c26c87e4197fe88ca883fa7e618a

**Review by alessandrod — APPROVED** [review]

**Review by behzadnouri — APPROVED** [review]

**Pull request merged**